### PR TITLE
Psalm plugin to squish mutations when they do not fit type declarations

### PR DESCRIPTION
**Pull request by Ocramius**

### CODE REVIEW

File: bin/roave-infection-static-analysis-plugin [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        define('PHP_PARSER_VERSION', Versions::getVersion('nikic/php-parser'));
+    }
+
+    $psalmConfig = Config::getConfigForPath($cwd, $cwd, Report::TYPE_CONSOLE);
+
+    $psalmConfig->setIncludeCollector(new IncludeCollector());
+
+    (new Application(Bootstrapper::bootstrap(
+        Container::create(),
+        new RunStaticAnalysisAgainstMutant($psalmConfig)
```
Comment (jdrieghe) [inline bin/roave-infection-static-analysis-plugin]: As this binary is psalm-specific, I would consider renaming this to `bin/roave-infection-psalm-plugin` to allow for a `bin/roave-infection-phpstan-plugin` alongside this one.

Comment (Ocramius) [reply bin/roave-infection-static-analysis-plugin]: That would come at a later point in time, when/if we do detection of which static analysis tool is installed (not `1.0.x`)

**Comment by Ocramius** [comment]
<details>
<summary>See https://github.com/vimeo/psalm/issues/4117</summary>
Tried this briefly on `ocramius/proxy-manager`, and it crashes when trying to find some symbols:

```
 ../../roave/infection-static-analysis-plugin/bin/roave-infection-static-analysis-plugin --test-framework-options='--testsuite=unit'  -vvv
[debug] Checking INFECTION_ALLOW_XDEBUG
[debug] The Xdebug extension is loaded (2.9.6)
[debug] Async signals enabled (SIGINT = SIG_IGN)
[debug] Process restarting (INFECTION_ALLOW_XDEBUG=internal|2.9.6|1|*|*)
[debug] Running '/usr/bin/php7.4' '../../roave/infection-static-analysis-plugin/bin/roave-infection-static-analysis-plugin' '--test-framework-options=--testsuite=unit' '-vvv' '--ansi'
[debug] Checking INFECTION_ALLOW_XDEBUG
[debug] Restarted (302 ms). The Xdebug extension is not loaded
[debug] Async signals enabled (SIGINT = SIG_DFL)

    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Infection - PHP Mutation Testing Framework version 0.17.5@14ff2578179588dfdc3cd05ceb8337cdedfdf5cc


Running initial test suite...

PHPUnit version: 9.2.5

  856 [============================] 5 secs

Generate mutants...

Processing source code files: 141/141
.: killed, M: escaped, U: uncovered, E: fatal error, T: timed out, S: skipped

....E.UU..M.......................................   ( 50 / 714)
...M.................
In Methods.php line 460:
                                                           
  [UnexpectedValueException]                               
  Cannot get method params for ReflectionClass::hasmethod  
                                                           

Exception trace:
  at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Methods.php:460
 Psalm\Internal\Codebase\Methods->getMethodParams() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/CallAnalyzer.php:255
 Psalm\Internal\Analyzer\Statements\Expression\CallAnalyzer::checkMethodArgs() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/Call/Method/AtomicMethodCallAnalyzer.php:670
 Psalm\Internal\Analyzer\Statements\Expression\Call\Method\AtomicMethodCallAnalyzer::analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/Call/MethodCallAnalyzer.php:173
 Psalm\Internal\Analyzer\Statements\Expression\Call\MethodCallAnalyzer::analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php:150
 Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php:45
 Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/BooleanNotAnalyzer.php:23
 Psalm\Internal\Analyzer\Statements\Expression\BooleanNotAnalyzer::analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php:273
 Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php:45
 Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Block/IfAnalyzer.php:603
 Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyzeIfConditional() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Block/IfAnalyzer.php:90
 Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php:448
 Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php:169
 Psalm\Internal\Analyzer\StatementsAnalyzer->analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php:597
 Psalm\Internal\Analyzer\FunctionLikeAnalyzer->analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php:1919
 Psalm\Internal\Analyzer\ClassAnalyzer->analyzeClassMethod() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php:748
 Psalm\Internal\Analyzer\ClassAnalyzer->analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php:221
 Psalm\Internal\Analyzer\FileAnalyzer->analyze() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php:357
 Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php:597
 Psalm\Internal\Codebase\Analyzer->doAnalysis() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php:283
 Psalm\Internal\Codebase\Analyzer->analyzeFiles() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php:1191
 Psalm\Internal\Analyzer\ProjectAnalyzer->checkFile() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/src/Roave/InfectionStaticAnalysis/Psalm/RunStaticAnalysisAgainstMutant.php:39
 Roave\InfectionStaticAnalysis\Psalm\RunStaticAnalysisAgainstMutant->isMutantStillValidAccordingToStaticAnalysis() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/src/Roave/InfectionStaticAnalysis/RunStaticAnalysisAgainstEscapedMutant.php:39
 Roave\InfectionStaticAnalysis\RunStaticAnalysisAgainstEscapedMutant->createFromProcess() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Process/Factory/MutantProcessFactory.php:101
 Infection\Process\Factory\MutantProcessFactory::Infection\Process\Factory\{closure}() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Process/MutantProcess.php:95
 Infection\Process\MutantProcess->terminateProcess() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Process/Runner/ParallelProcessRunner.php:125
 Infection\Process\Runner\ParallelProcessRunner->freeTerminatedProcesses() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Process/Runner/ParallelProcessRunner.php:98
 Infection\Process\Runner\ParallelProcessRunner->run() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Process/Runner/MutationTestingRunner.php:126
 Infection\Process\Runner\MutationTestingRunner->run() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Engine.php:166
 Infection\Engine->runMutationAnalysis() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Engine.php:110
 Infection\Engine->execute() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Command/RunCommand.php:240
 Infection\Command\RunCommand->executeCommand() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/infection/infection/src/Command/BaseCommand.php:78
 Infection\Command\BaseCommand->execute() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/symfony/console/Command/Command.php:258
 Symfony\Component\Console\Command\Command->run() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/symfony/console/Application.php:916
 Symfony\Component\Console\Application->doRunCommand() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/symfony/console/Application.php:264
 Symfony\Component\Console\Application->doRun() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/vendor/symfony/console/Application.php:140
 Symfony\Component\Console\Application->run() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/bin/roave-infection-static-analysis-plugin:46
 Roave\InfectionStaticAnalysis\Application\{closure}() at /home/ocramius/Documents/roave/infection-static-analysis-plugin/bin/roave-infection-static-analysis-plugin:47

run [--test-framework TEST-FRAMEWORK] [--test-framework-options TEST-FRAMEWORK-OPTIONS] [-j|--threads THREADS] [--only-covered] [-s|--show-mutations] [--no-progress] [--force-progress] [-c|--configuration CONFIGURATION] [--coverage COVERAGE] [--mutators MUTATORS] [--filter FILTER] [--formatter FORMATTER] [--min-msi MIN-MSI] [--min-covered-msi MIN-COVERED-MSI] [--log-verbosity LOG-VERBOSITY] [--initial-tests-php-options INITIAL-TESTS-PHP-OPTIONS] [--skip-initial-tests] [--ignore-msi-with-no-mutations] [--debug] [--dry-run]

[debug] Restarted process exited 1
```

@muglug can I ask for some help in getting the `Configuration` object set up correctly? Specifically looking at this bit: https://github.com/Roave/infection-static-analysis-plugin/pull/1/files#diff-ea8ef2c40e7c5062074978631316bd99R35-R37

I'm not sure if I'm setting up `vimeo/psalm` correctly :thinking: 
</details>

**Comment by Ocramius** [comment]
With 79819d65cbb6c74acc4f40d7af517bdce4beabd2, `./vendor/bin/phpunit --random-order-seed=1599077027` works, `./vendor/bin/phpunit --random-order-seed=1599077026` doesn't

**Comment by Ocramius** [comment]
@asgrim with your review, I'd call this a `1.0.0` :+1:

**Comment by Ocramius** [comment]
Tested on [`laminas/automatic-releases`](https://github.com/laminas/automatic-releases):

 * Mutation Score Indicator (MSI): 64%
 * Mutation Code Coverage: 97%
 * Covered Code MSI: 66%

Becomes:

 * Mutation Score Indicator (MSI): 75%
 * Mutation Code Coverage: 97%
 * Covered Code MSI: 77%

File: .gitattributes [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/test export-ignore
+.gitattributes export-ignore
```
Comment (Ocramius) [inline .gitattributes:3]: Should add `.github` here

File: infection.json.dist [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+    "source": {
+        "directories": [
+            "src"
+        ]
+    },
+    "logs": {
+        "text": "php://stderr",
+        "badge": {
+            "branch": "master"
```
Comment (Ocramius) [inline infection.json.dist:10]: This needs setting up env var secrets

File: src/Roave/InfectionStaticAnalysis/Bootstrapper.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+/** @internal */
+final class Bootstrapper
+{
+    public static function bootstrap(
+        Container $container,
+        RunStaticAnalysisAgainstMutant $runStaticAnalysis
+    ): Container {
+        $reflectionOffsetSet = (new ReflectionMethod(Container::class, 'offsetSet'));
+
+        // @TODO we may want to make this one lazy, but for now this is OK
```
Comment (Ocramius) [inline src/Roave/InfectionStaticAnalysis/Bootstrapper.php:21]: Can't really make it lazy for now, but may be improved at a later point in time

File: .gitattributes (hunk 2 above)
Comment (Ocramius) [reply .gitattributes:3]: 0121d0c112b785d1eeca01953b67ea2e03cb6de7

File: infection.json.dist (hunk 3 above)
Comment (Ocramius) [reply infection.json.dist:10]: Configured. Won't work for a while, until branch matchers are supported :-)

File: README.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# Infection Static Analysis Plugin
+
+This plugin is designed to run static analysis on top of [`infection/infection`](https://github.com/infection/infection)
+test runs in order to discover if [escaped mutants](https://en.wikipedia.org/wiki/Mutation_testing)
+are valid mutations, or if they do not respect the type signature of your
+program.
```
Comment (asgrim) [inline README.md]:
```suggestion
program. If the mutation would result a type error, it is "killed".

TL;DR:

 - This will improves your mutation score, since mutations which result in type errors become killed.
 - This is very hacky, and replaces `vendor/bin/infection` essentially. Please read the `Stability` section below first for details.
 - This is currently much slower than running infection by itself. There are ideas/suggestions to improve this in the future.
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+## Usage
+
+```sh
+composer require roave/infection-static-analysis-plugin
```
Comment (asgrim) [inline README.md]:
```suggestion
composer require --dev roave/infection-static-analysis-plugin
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
+# Infection Static Analysis Plugin
+
+This plugin is designed to run static analysis on top of [`infection/infection`](https://github.com/infection/infection)
+test runs in order to discover if [escaped mutants](https://en.wikipedia.org/wiki/Mutation_testing)
+are valid mutations, or if they do not respect the type signature of your
+program.
+
+## Usage
+
```
Comment (asgrim) [inline README.md:17]: The current design of this tool is to run `vendor/bin/roave-infection-static-analysis-plugin` instead of running `vendor/bin/infection`.

File: README.md (hunk 5 above)
Comment (Ocramius) [reply README.md]: Is it "would result **in** a type error"? Asking because I never know whether to add `in`

Comment (Ocramius) [reply README.md]: b7b501460592ecacbce781c8dc0e383917dd06fa

File: README.md (hunk 7 above)
Comment (Ocramius) [reply README.md:17]: b7b501460592ecacbce781c8dc0e383917dd06fa

File: README.md (hunk 6 above)
Comment (Ocramius) [reply README.md]: b7b501460592ecacbce781c8dc0e383917dd06fa

File: README.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
+# Infection Static Analysis Plugin
+
+This plugin is designed to run static analysis on top of [`infection/infection`](https://github.com/infection/infection)
+test runs in order to discover if [escaped mutants](https://en.wikipedia.org/wiki/Mutation_testing)
+are valid mutations, or if they do not respect the type signature of your
+program. If the mutation would result in a type error, it is "killed".
+
+TL;DR:
+ - This will improves your mutation score, since mutations which result in
```
Comment (asgrim) [inline README.md]:
```suggestion
 - This will improve your mutation score, since mutations which result in
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+   type errors become killed.
+ - This is very hacky, and replaces `vendor/bin/infection` essentially.
+   Please read the `Stability` section below first for details.
+ - This is currently much slower than running infection by itself.
+   There are ideas/suggestions to improve this in the future.
+ 
+## Usage
+
+The current design of this tool requires to run `vendor/bin/roave-infection-static-analysis-plugin`
```
Comment (asgrim) [inline README.md]:
```suggestion
The current design of this tool requires you to run `vendor/bin/roave-infection-static-analysis-plugin`
```

**Review by asgrim — APPROVED** [review]

**Comment by Ocramius** [comment]
:ship: as per discussion with @asgrim
